File: internal/agent/ambient_state_test.go
package agent

import (
	"encoding/json"
	"strings"
	"testing"

	"github.com/camronwood/neural-junkie/internal/protocol"
)

func ambientTestMessage(content string, state interface{}) *protocol.Message {
	return &protocol.Message{
		Content:  content,
		Metadata: map[string]interface{}{MetadataAmbientState: state},
	}
}

func TestSanitizeAmbientStateMetadataRelevanceRedactionAndCap(t *testing.T) {
	diagnostics := make([]interface{}, 0, 100)
	for i := 0; i < 100; i++ {
		diagnostics = append(diagnostics, map[string]interface{}{
			"path": "/repo/main.go", "line": i + 1, "column": 1, "severity": "error",
			"message": "password=hunter2 " + strings.Repeat("x", 1000),
		})
	}
	msg := ambientTestMessage("fix the diagnostics", map[string]interface{}{
		"active_editor": map[string]interface{}{
			"path":      "/repo/.env",
			"selection": map[string]interface{}{"start_line": 1, "end_line": 1, "text": "api_key=secret"},
		},
		"diagnostics": diagnostics,
		"terminal": map[string]interface{}{
			"cwd": "/repo", "failed_tail": "\x1b[31merror client_secret=secret\x1b[0m",
		},
	})
	SanitizeAmbientStateMetadata(msg)

	state, ok := msg.Metadata[MetadataAmbientState].(map[string]interface{})
	if !ok {
		t.Fatal("expected sanitized ambient state")
	}
	encoded, _ := json.Marshal(state)
	if len(encoded) > maxAmbientStateServerBytes {
		t.Fatalf("ambient state is %d bytes, cap is %d", len(encoded), maxAmbientStateServerBytes)
	}
	if strings.Contains(string(encoded), "hunter2") || strings.Contains(string(encoded), "client_secret=secret") || strings.Contains(string(encoded), "\x1b") {
		t.Fatalf("unsafe text remained: %s", encoded)
	}
	editor := state["active_editor"].(map[string]interface{})
	selection := editor["selection"].(map[string]interface{})
	if _, exists := selection["text"]; exists {
		t.Fatal("sensitive-file selection must be path/coordinates only")
	}
}

func TestSanitizeAmbientStateMetadataDropsIrrelevantState(t *testing.T) {
	msg := ambientTestMessage("What is the weather?", map[string]interface{}{
		"terminal": map[string]interface{}{"failed_tail": "error"},
	})
	SanitizeAmbientStateMetadata(msg)
	if _, exists := msg.Metadata[MetadataAmbientState]; exists {
		t.Fatal("irrelevant ambient state should be removed")
	}
}

func TestAppendAmbientStateRendersEphemeralSection(t *testing.T) {
	msg := ambientTestMessage("debug this", map[string]interface{}{
		"recent_edits": []interface{}{map[string]interface{}{"path": "/repo/main.go", "edited_at": 1}},
	})
	SanitizeAmbientStateMetadata(msg)
	var prompt strings.Builder
	AppendAmbientState(&prompt, msg)
	if !strings.Contains(prompt.String(), "AMBIENT IDE STATE (EPHEMERAL)") ||
		!strings.Contains(prompt.String(), "/repo/main.go") {
		t.Fatalf("unexpected prompt section: %q", prompt.String())
	}
}

File: desktop/src/utils/ambientState.test.ts
import { beforeEach, describe, expect, it } from 'vitest';
import {
  CLIENT_AMBIENT_STATE_TARGET_BYTES,
  ambientStateIsRelevant,
  buildAmbientState,
  sanitizeAmbientText,
} from './ambientState';
import { useDiagnosticsStore } from '../stores/diagnosticsStore';
import { useEditorStore } from '../stores/editorStore';
import { useFileExplorerStore } from '../stores/fileExplorerStore';
import { useTerminalStore } from '../stores/terminalStore';

describe('ambient IDE state', () => {
  beforeEach(() => {
    useEditorStore.setState({
      tabs: [
        {
          id: 'active',
          workspaceId: 'ws',
          path: '/repo/src/main.ts',
          content: 'const value = 1;',
          language: 'typescript',
          isDirty: true,
          cursorPosition: { line: 1, column: 6 },
        },
      ],
      activeTabId: 'active',
      activeSelection: {
        tabId: 'active',
        startLine: 1,
        endLine: 1,
        text: 'api_key=super-secret',
      },
      recentEdits: [{ path: '/repo/src/main.ts', editedAt: 123 }],
    });
    useFileExplorerStore.setState({
      workspaces: [{ id: 'ws', name: 'repo', path: '/repo', kind: 'local', is_git_repo: true }],
      activeWorkspaceId: 'ws',
    });
    useDiagnosticsStore.setState({
      byPath: {
        '/repo/src/main.ts': [
          { path: '/repo/src/main.ts', line: 1, column: 1, message: 'password=hunter2', severity: 'error' },
        ],
      },
    });
    useTerminalStore.setState({
      activeTabId: 'tab-0',
      recentFailedTails: { 'tab-0': '\u001b[31merror: token=abc123\u001b[0m' },
    });
  });

  it('does not infer ambient relevance from natural-language phrases', () => {
    expect(ambientStateIsRelevant('How is the weather?')).toBe(false);
    expect(ambientStateIsRelevant('Fix the current editor error')).toBe(false);
    expect(buildAmbientState('Fix the current editor error')).toBeUndefined();
  });

  it('assembles and redacts when structurally forced (IDE / context_request)', () => {
    expect(ambientStateIsRelevant('anything', true)).toBe(true);
    const state = buildAmbientState('ignored', { ideCoding: true });
    expect(state?.active_editor?.selection?.text).toContain('api_key=[REDACTED]');
    expect(state?.diagnostics?.[0]?.message).toContain('password=[REDACTED]');
    expect(state?.terminal?.failed_tail).not.toContain('\u001b');
    expect(state?.recent_edits).toEqual([{ path: '/repo/src/main.ts', edited_at: 123 }]);
  });

  it('does not attach selection text to a different or sensitive active tab', () => {
    useEditorStore.setState((state) => ({
      tabs: [{ ...state.tabs[0], path: '/repo/.env' }],
      activeSelection: { ...state.activeSelection!, tabId: 'stale-tab' },
    }));
    expect(buildAmbientState('debug this file', { ideCoding: true })?.active_editor?.selection).toBeUndefined();

    useEditorStore.setState({
      activeSelection: { tabId: 'active', startLine: 1, endLine: 1, text: 'SECRET=visible' },
    });
    expect(buildAmbientState('debug this file', { ideCoding: true })?.active_editor?.selection?.text).toBeUndefined();
  });

  it('stays within the 8KB client target', () => {
    useDiagnosticsStore.setState({
      byPath: {
        '/repo/src/main.ts': Array.from({ length: 200 }, (_, i) => ({
          path: '/repo/src/main.ts',
          line: i + 1,
          column: 1,
          message: `error ${i} `.repeat(20),
          severity: 'error' as const,
        })),
      },
    });
    const state = buildAmbientState('ignored', { ideCoding: true, force: true });
    expect(new TextEncoder().encode(JSON.stringify(state)).length).toBeLessThanOrEqual(
      CLIENT_AMBIENT_STATE_TARGET_BYTES + 2048,
    );
  });

  it('sanitizes ambient text helpers', () => {
    expect(sanitizeAmbientText('password=hunter2')).toContain('[REDACTED]');
  });
});
